### ISSUE TITLE
Create a bidirectional conversion between StableHLO and TCP

### ISSUE THREAD

**Issue by burmako** [openxla/stablehlo#17]
TCP dialect is a super exciting initiative in MLIR upstream ([Discourse thread](https://discourse.llvm.org/t/development-of-high-level-tensor-compute-primitives-dialect-s-and-transformations/388), [design document](https://docs.google.com/document/d/1f3KVsXA4xm6W7gd2cKx9ThGA52XMfZKsPV3ZOlsztC4/edit#))! Let's contribute to this effort by providing converters to/from TCP, so that TCP work can potentially leverage them to connect with the existing MHLO and, in the future, StableHLO producers (which currently include JAX, PyTorch and TensorFlow).

**Comment by burmako** [comment]
Thank you! Looks like I accidentally copy/pasted the wrong link from the design document :)

**Comment by burmako** [comment]
Current status: there is a proposal to add TCP to the LLVM Incubator ([Discourse post](https://discourse.llvm.org/t/rfc-proposal-for-a-high-level-ml-dialect-in-mlir/64249/142), and there'll be an MLIR open design meeting to discuss the proposed spec next Thursday 9/1 ([agenda](https://docs.google.com/document/d/1y_9f1AbfgcoVdJh4_aM6-BaSHvrHl8zuA5G4jv_94K8/edit)). Nothing actionable for the StableHLO side just, but we're really looking forward to further progress!

**Comment by burmako** [comment]
Great progress on the TCP side which now has [a dedicated category on Discourse](https://discourse.llvm.org/c/mlir/mlir-tcp-wg/36) and [an initial draft of the spec](https://discourse.llvm.org/t/rfc-initial-draft-of-tcp-spec/64973) which we discussed at an MLIR ODM last Thursday 9/1. 

It looks like a code repository with the TCP dialect is going to be created very soon, and I'm looking forward to starting working on the conversion once it's there! 

I've also looked into the logistics of integrating a potential GitHub-first dependency from StableHLO to TCP into the Google-first MLIR-HLO repository. On the technical side, the machinery is fairly straightforward. On the organizational side, this will depend on the eventual license that TCP adopts.

**Comment by burmako** [comment]
Exciting news! TCP is being bootstrapped in Torch-MLIR (https://github.com/llvm/torch-mlir/issues/1366), and there's a pull request that introduces the needed boilerplate for the TCP dialect as well as a few ops - `tanh`, `add` and `matmul` (https://github.com/llvm/torch-mlir/pull/1375).

As far as conversions between StableHLO and TCP go, the current plan is to wait until the above mentioned pull request lands and then start developing the conversions in Torch-MLIR. Torch-MLIR depends on MLIR-HLO, and StableHLO is shipped as part of MLIR-HLO, so we're already all set dependency-wise. I can't wait!

**Comment by bhack** [comment]
Do you mean that initially you will work on this conversion inside the torch-mlir incubator repo?

I suppose this cause I don't think we could create a dependency here between this project and torch-mlir repo.

**Comment by burmako** [comment]
https://github.com/llvm/torch-mlir/pull/1375 has just landed in the Torch-MLIR repository, creating a TCP dialect and setting up all the related boilerplate + a `tcp.dummy` op. I'm extremely excited for the future development that this PR unlocks! 👏

**Comment by burmako** [comment]
@navahgar What next steps for TCP do you have in mind? Anything we need to prepare to on the StableHLO side? In the [the SIG MLIR Open Design Meeting document](https://docs.google.com/document/d/1y_9f1AbfgcoVdJh4_aM6-BaSHvrHl8zuA5G4jv_94K8/edit#heading=h.cite1kolful9), I saw "Broadcasting op & broadcasting semantics in TCP (Raghavan Raman, Cruise)", and I'm looking forward to the ODM :)

**Comment by navahgar** [comment]
@burmako Our plan is to get to the elementwise ops in TCP first. But that requires broadcasting, which in turn touches several important design points. We are hoping to get through with a decision on broadcasting and elementwise ops soon, preferably in the ODM.

**Comment by burmako** [comment]
Here are the corresponding design documents: [Broadcast](https://docs.google.com/document/d/1x8kCUsYzFEIeWO38h32bHlmDrieeu9UJTxpKe9gUm_g/edit#heading=h.ey5ld3h9qkk2) and [Elementwise](https://docs.google.com/document/d/1fP5qOosnA2kVqX50g6r5DwDnSbOBAcrcPFOsL7RAbcQ/edit#heading=h.41d1zukuh2n7). We've discussed the former at the ODM last Thursday 10/13 and came to an intermediate resolution which will be followed by prototyping. Elementwise proposal still needs to be discussed.

**Comment by bhack** [comment]
Is TCP still an active upstream project? As It seems to me not.

**Comment by sjarus** [comment]
> Is TCP still an active upstream project? As It seems to me not.

It's very much in active development - we'd shifted focus to internally prototyping some of its important constructs and transformations for use in our deployment flows, and to enable us to provide feedback from that experience publicly. We'll be sharing some of this through additional design documents and code in the near future.

**Comment by bhack** [comment]
Ok thanks for the clarification as I saw no more public activities on Github.

**Comment by burmako** [comment]
https://github.com/llvm/torch-mlir/pull/1932 introduces the boilerplate for the StableHLO => TCP pass as well as an implementation of converting stablehlo.tanh to tcp.tanh to make sure that all the boilerplate works. Once we're aligned on logistics, I'll implement the rest :)

**Comment by bhack** [comment]
Is TCP actively maintained?

https://github.com/llvm/torch-mlir/issues/1366

**Comment by navahgar** [comment]
> Is TCP actively maintained?

Yes. This branch all the code: https://github.com/llvm/torch-mlir/tree/mlir-tcp

We have also been posting PRs to the discord channel : https://discord.com/channels/636084430946959380/1020532515720671332
